### PR TITLE
Whisper: fix static cache CI

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Some tests started failing after https://github.com/huggingface/transformers/pull/35164. This PR fixes it by:
1. Removing manual forward compilation, as it is compiled again in the generation's `_sample`
2. Removes the overriden `prepare_inputs_for_generation` 

Ran tests:
- `pytest -k greedy tests/models` (moshi is failing even on `main`)
- `pytest  tests/models/whisper/test_modeling_whisper.py`

### CODE REVIEW

File: tests/models/whisper/test_modeling_whisper.py [hunk 1]
```diff
@@ -3318,8 +3318,8 @@ def test_tiny_static_generation(self):
         input_features = input_features.to(torch_device)
         eager_generated_ids = model.generate(input_features, max_new_tokens=64)
 
+        # Using statiic cache compiles forward for each decoding step, so we don't have to manually compile
```
Comment (ArthurZucker) [inline tests/models/whisper/test_modeling_whisper.py:3321]: that is only true for normal models, but not sure for whisper no? as it overwrites `generate`

**Review by ArthurZucker — APPROVED** [review]
Nice that the slack ping worked!

File: tests/models/whisper/test_modeling_whisper.py (hunk 1 above)
Comment (zucchini-nlp) [reply tests/models/whisper/test_modeling_whisper.py:3321]: same as Whisper call `super().generate()` after model-specific input preparations

Comment (ArthurZucker) [reply tests/models/whisper/test_modeling_whisper.py:3321]: Okay, thanks!˜

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -406,23 +406,28 @@ def prepare_inputs_for_generation(
...
+            and position_ids_key in set(inspect.signature(self.forward).parameters.keys())
         ):
             position_ids = attention_mask.long().cumsum(-1) - 1
             position_ids.masked_fill_(attention_mask == 0, 1)
-            kwargs["position_ids"] = position_ids  # placed in kwargs for further processing (see below)
+            kwargs[position_ids_key] = position_ids  # placed in kwargs for further processing (see below)
 
         # 5. Slice model inputs if it's an input that should have the same length as `input_ids`
-        for model_input_name in ["position_ids", "token_type_ids"]:
+        for model_input_name in ["position_ids", "token_type_ids", "decoder_position_ids"]:
```
Comment (gante) [inline src/transformers/generation/utils.py:424]: `input_ids`, the argument, should contain the data (if any) from `decoder_position_ids` in encoder-decoder models (see `_prepare_decoder_input_ids_for_generation`). As such, we should discard `decoder_position_ids` here.

(or am I missing anything?)

File: src/transformers/models/whisper/generation_whisper.py [hunk 3]
```diff
@@ -1234,7 +1234,7 @@ def _expand_variables_for_generation(
     def _setup_no_speech_detection(logits_processor, segment_input, decoder_input_ids, kwargs):
         set_inputs = _get_attr_from_logit_processors(logits_processor, WhisperNoSpeechDetection, "set_inputs")
         extra_kwargs = {k: v for k, v in kwargs.items() if torch.is_tensor(v)}
-        set_inputs({"inputs": segment_input, "decoder_input_ids": decoder_input_ids, **extra_kwargs})
+        set_inputs({"inputs": segment_input, "input_ids": decoder_input_ids, **extra_kwargs})
```
Comment (gante) [inline src/transformers/models/whisper/generation_whisper.py:1237]: if this is for the input of the generalist `generate`, then this change shouldn't be needed (see comment above)

File: tests/models/whisper/test_modeling_whisper.py (hunk 1 above)
Comment (gante) [inline tests/models/whisper/test_modeling_whisper.py:3321]:
```suggestion
        # Using static cache compiles forward for each decoding step, so we don't have to manually compile
```

**Review by gante — COMMENTED** [review]
In general LGTM, but I have a question

File: src/transformers/models/whisper/generation_whisper.py (hunk 3 above)
Comment (zucchini-nlp) [reply src/transformers/models/whisper/generation_whisper.py:1237]: Without this change, the `input_ids` will be `None` because we pass all keyword args from here, and `input_ids` is not in the keywords

File: src/transformers/generation/utils.py (hunk 2 above)
Comment (zucchini-nlp) [reply src/transformers/generation/utils.py:424]: I see, `_prepare_decoder_input_ids_for_generation` actually does nothing about position ids or taking care of `decoder_` prefix to args. It only looks prepares `input_ids` which means we'll need to manually prepare and pass other args here   

- `decoder_position_ids`: needed because otherwise we will be trying to infer it from `cache_position.unsqueeze(0)` and not taking into account the masking. Also needed for `compile` as we need cloned position ids each step, not the ones inferred from cache position
- `decoder_attention_mask`: needed because we prepare position ids based on attention mask length which has to be a decoder mask and same reason as above. We need the model to mask out attention weights correctly 
- `input_ids_key`: added so we can easily get input length and crop extra kwargs to same length, instead of checking for `None` every time

File: src/transformers/models/whisper/generation_whisper.py (hunk 3 above)
Comment (gante) [reply src/transformers/models/whisper/generation_whisper.py:1237]: ah, I see

**Review by gante — APPROVED** [review]

**Comment by zucchini-nlp** [comment]
Merging after tests are fixed on `main`
